hazelcast/include/hazelcast/client/Socket.h

```diff
@@ -16,67 +16,44 @@
 #ifndef HAZELCAST_CLIENT_SOCKET_H_
 #define HAZELCAST_CLIENT_SOCKET_H_
 
-#include <memory>
-#include <memory>
+#include <boost/asio.hpp>
 
+#include "hazelcast/client/config/SocketOptions.h"
 #include "hazelcast/client/Address.h"
 
 #if  defined(WIN32) || defined(_WIN32) || defined(WIN64) || defined(_WIN64)
 #pragma warning(push)
 #pragma warning(disable: 4251) //for dll export	
-#endif 
+#endif
 
 namespace hazelcast {
     namespace client {
-        /**
-         * Sockets wrapper interface class.
-         */
+        namespace connection {
+            class Connection;
+
+            class AuthenticationFuture;
+        }
+
+        namespace protocol {
+            class ClientMessage;
+        }
+
         class HAZELCAST_API Socket {
         public:
-            /**
-             * Destructor
-             */
-            virtual ~Socket();
-            /**
-             * connects to given address in constructor.
-             * @param timeoutInMillis if not connected within timeout, it will return errorCode
-             * @return 0 on success.
-             * @throw IOException on failure.
-             */
-            virtual int connect(int timeoutInMillis) = 0;
+            virtual ~Socket() {
+            }
 
-            /**
-             * @param buffer
-             * @param len length of the buffer
-             * @param flag bsd sockets options flag. Only MSG_WAITALL is supported when SSL is enabled.
-             * @return number of bytes send
-             * @throw IOException in failure.
-             */
-            virtual int send(const void *buffer, int len, int flag = 0) = 0;
+            virtual void
+            asyncStart(const std::shared_ptr<connection::Connection> &connection,
```

Comment:
It seems like the levels of concerns are mixed. A socket should not now about connection or authentication. I think this is not a socket anymore. 

I only see one implementation of this which is named BaseSocket, which also seems confusing. It seems that there is no need for this class to be pure virtual anymore.

Reply:
Discussed the reasons for this. Will leave it this way for now. In Java client master code these also changed, hence will change this again in the future.